File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/DP/LargestPalindromicSubstr.swift
import Foundation
// Palindromic Substrings

func largestPalinSubstr(_ str : String) -> Int {
	let n : Int = str.count
	var dp : [[Int]] = Array(repeating: Array(repeating: 0, count: n), count: n)
	var i : Int = 0
	while (i < n) {
		dp[i][i] = 1
		i += 1
	}
	
	var max : Int = 1
	var start : Int = 0
	var l : Int = 1
	while (l < n) {
		var i : Int = 0, j : Int = i + l
		while (j < n) {
			if (Array(str)[i] == Array(str)[j] && dp[i + 1][j - 1] == j - i - 1) {
				dp[i][j] = dp[i + 1][j - 1] + 2
				if (dp[i][j] > max) {
					max = dp[i][j]
					// Keeping track of max length and
					start = i
				}
			} else {
				dp[i][j] = 0
			}
			i += 1
			j += 1
		}
		l += 1
	}
	let str2 = (str.prefix(start + max)).suffix(max)
	print("Max Length Palindromic Substrings : " + str2)
	return max
}

// Testing code.
let str : String = "ABCAUCBCxxCBA"
print("Max Palindromic Substrings len:", largestPalinSubstr(str))

/* 
Max Length Palindromic Substrings : BCxxCB
Max Palindromic Substrings len: 6
 */

File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/DP/LargestPalindromicSubstr.ts
function largestPalinSubstr(str: string): number {
	const n = str.length;
	const dp: number[][] = new Array(n);

	for (let i = 0; i < n; i++) {
		dp[i] = new Array(n).fill(0);
		dp[i][i] = 1;
	}

	let max = 1;
	let start = 0;
	for (let l = 1; l < n; l++) {
		for (let i = 0, j = i + l; j < n; i++, j++) {
			if (str.charAt(i) === str.charAt(j) && dp[i + 1][j - 1] === j - i - 1) {
				dp[i][j] = dp[i + 1][j - 1] + 2;
				if (dp[i][j] > max) {
					max = dp[i][j];
					start = i;
				}
			} else {
				dp[i][j] = 0;
			}
		}
	}
	console.log("Max Length Palindromic Substrings:", str.substring(start, start + max));
	return max;
}

const str = "ABCAUCBCxxCBA";
console.log("Max Palindromic Substrings len:", largestPalinSubstr(str));


/*
Max Length Palindromic Substrings: BCxxCB
Max Palindromic Substrings len: 6
*/

/*
 * If asked to find how many different palindromic substrings are possible.
 * 
 * int count = 0; 
 * for(int i=0;i<n;i++) 
 * 	for(int j=0;j<n;j++) 
 * 		if(dp[i][j] > 0)
 * 			count++; return count;
 */
